modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpPipeliningHandler.java

```diff
@@ -324,7 +324,6 @@ private boolean writeChunk(ChannelHandlerContext ctx, PromiseCombiner combiner,
             Netty4WriteThrottlingHandler.MAX_BYTES_PER_WRITE,
             serverTransport.recycler()
         );
-        assert bytes.length() > 0 : "serialization should not produce empty buffers";
```

Comment:
IIUC, this assertion ensures two things:
1. Each invocation of `encodeChunk` return non-empty buffer
2. Because of 1, the overall buffer written for the response body is non-empty.

The test proves that we want to remove the first item, i.e. there can be empty buffers produced in the middle of writing the entire reponse. But the test does _not_ suggest that we should remove item 2, i.e. the overall test response is in fact non-empty. I am not sure whether we want to honor item 2 here or is the intention to drop it as well? If we want to drop it, I suggest the test response should sometimes be entirely empty. Otherwise, it seems that we need a new assertion somewhere else, e.g. call site of this method?

Reply:
Fair point, I'll add a test for that case too. It should indeed be fine, and acceptable, to send an empty response.